lib/Driver/ToolChains.cpp

```diff
@@ -1228,19 +1228,21 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
 
   addInputsOfType(Arguments, context.InputActions, types::TY_Object);
 
-  if (context.OI.DebugInfoKind > IRGenDebugInfoKind::LineTables) {
-    size_t argCount = Arguments.size();
-    if (context.OI.CompilerMode == OutputInfo::Mode::SingleCompile)
-      addInputsOfType(Arguments, context.Inputs, types::TY_SwiftModuleFile);
-    else
-      addPrimaryInputsOfType(Arguments, context.Inputs,
-                             types::TY_SwiftModuleFile);
+  size_t argCount = Arguments.size();
+  if (context.OI.CompilerMode == OutputInfo::Mode::SingleCompile) {
+    addInputsOfType(Arguments, context.Inputs, types::TY_SwiftModuleFile);
+  } else {
+    addPrimaryInputsOfType(Arguments, context.Inputs,
+                           types::TY_SwiftModuleFile);
+  }
 
-    if (Arguments.size() > argCount) {
-      assert(argCount + 1 == Arguments.size() &&
-             "multiple swiftmodules found for -g");
-      Arguments.insert(Arguments.end() - 1, "-add_ast_path");
-    }
+  addInputsOfType(Arguments, context.InputActions, types::TY_SwiftModuleFile);
+
+  // For each .swiftmodule file added to the linker input arguments, prepend
+  // an -add_ast_path option.
```

Comment:
This seems a bit too clever. What do you think about adding an extra `const char *` argument to `addInputsOfType` that is inserted before each input if present?